simple/rdm_shared_av.c

```diff
@@ -0,0 +1,237 @@
+/*
+ * Copyright (c) 2013-2015 Intel Corporation.  All rights reserved.
+ *
+ * This software is available to you under the BSD license
+ * below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AWV
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <getopt.h>
+#include <unistd.h>
+#include <sys/types.h>
+#include <sys/wait.h>
+#include <string.h>
+
+#include <rdma/fi_errno.h>
+#include <rdma/fi_endpoint.h>
+#include <rdma/fi_cm.h>
+
+#include "shared.h"
+
+static int parent;
+static int pair[2];
+static int ready;
+
+static int init_fabric(void)
+{
+	char *node, *service;
+	uint64_t flags = 0;
+	int ret;
+
+	ret = ft_read_addr_opts(&node, &service, hints, &flags, &opts);
+	if (ret)
+		return ret;
+
+	/* Get fabric info */
+	ret = fi_getinfo(FT_FIVERSION, node, service, flags, hints, &fi);
+	if (ret) {
+		FT_PRINTERR("fi_getinfo", ret);
+		return ret;
+	}
+
+	ret = ft_open_fabric_res();
+	if (ret)
+		return ret;
+
+	if (opts.dst_addr && !parent) {
+		/* child waits until parent is done creating AV */
+		ret = read(pair[0], &ready, sizeof(int));
+		if (ret < 0)
+			FT_PRINTERR("read", errno);
+
+		/* child needs to open AV in read only mode */
+		av_attr.flags = FI_READ;
+	}
+	ret = ft_alloc_active_res(fi);
+	if (ret)
+		return ret;
+
+	if (opts.dst_addr && parent) {
+		/* parent lets child know that AV is created */
+		ret = write(pair[1], &ready, sizeof(int));
+		if (ret < 0)
+			FT_PRINTERR("write", errno);
+	}
+
+	ret = ft_init_ep();
```

Comment:
This should probably move above the write() call to ensure that the server is ready to receive a message before the client tries to send it